libpromises/evalfunction.c

```diff
@@ -4429,70 +4429,123 @@ static FnCallResult FnCallFold(EvalContext *ctx,
+    if (!ref)
+    {
+        // no such variable
+        return FnFailure();
+    }
+
     const void *value = EvalContextVariableGet(ctx, ref, &type);
     VarRefDestroy(ref);
 
-    Writer* const typestring = StringWriter();
+    const char* major = NULL;
+    const char* minor = NULL;
+
+    const bool check_mode = (0 == strcmp(fp->name, "is_datatype"));
 
     if (type == CF_DATA_TYPE_CONTAINER)
     {
-
         const JsonElement* const jelement = value;
+        major = DataTypeToString(type);
 
         if (JsonGetElementType(jelement) == JSON_ELEMENT_TYPE_CONTAINER)
         {
             switch (JsonGetContainerType(jelement))
             {
             case JSON_CONTAINER_TYPE_OBJECT:
-                WriterWrite(typestring, "json_object");
+                minor = "object";
                 break;
             case JSON_CONTAINER_TYPE_ARRAY:
-                WriterWrite(typestring, "json_array");
+                minor = "array";
                 break;
             }
         }
         else if (JsonGetElementType(jelement) == JSON_ELEMENT_TYPE_PRIMITIVE)
         {
             switch (JsonGetPrimitiveType(jelement))
             {
-            case JSON_PRIMITIVE_TYPE_STRING:
-                WriterWrite(typestring, "json_string");
-                break;
-            case JSON_PRIMITIVE_TYPE_INTEGER:
-                WriterWrite(typestring, "json_integer");
-                break;
-            case JSON_PRIMITIVE_TYPE_REAL:
-                WriterWrite(typestring, "json_real");
-                break;
-            case JSON_PRIMITIVE_TYPE_BOOL:
-                WriterWrite(typestring, "json_bool");
-                break;
             case JSON_PRIMITIVE_TYPE_NULL:
-                WriterWrite(typestring, "json_null");
+                minor = "null";
+                break;
+            default:
+                minor = JsonPrimitiveTypeToString(JsonGetPrimitiveType(jelement));
                 break;
             }
         }
 
     }
-    else
+    else if (type == CF_DATA_TYPE_STRING ||
+             type == CF_DATA_TYPE_INT ||
+             type == CF_DATA_TYPE_REAL)
+    {
+        major = "primitive";
+        minor = DataTypeToString(type);
+    }
+    else if (type == CF_DATA_TYPE_STRING_LIST ||
+             type == CF_DATA_TYPE_INT_LIST ||
+             type == CF_DATA_TYPE_REAL_LIST)
     {
-        Log(LOG_LEVEL_VERBOSE, "%s: variable '%s' is not a data container", fp->name, varname);
+        major = "list";
+        minor = DataTypeToString(type);
+    }
+    else if (type == CF_DATA_TYPE_INT_RANGE ||
+             type == CF_DATA_TYPE_REAL_RANGE)
+    {
+        major = "range";
+        minor = DataTypeToString(type);
+    }
+
+    if (NULL == major || NULL == minor)
+    {
+        Log(LOG_LEVEL_VERBOSE, "%s: the data type of variable '%s' could not be found", fp->name, varname);
         return FnFailure();
     }
 
-    return FnReturnNoCopy(StringWriterClose(typestring));
+    if (check_mode)
```

Comment:
This is not reusing any part of the function, so why nest it in here and not create a new one?